internals/overlord/pairingstate/manager.go

```diff
@@ -0,0 +1,277 @@
+// Copyright (c) 2025 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// pairingstate manages client server pairing.
+package pairingstate
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/plan"
+)
+
+// timeAfterFunc allows faking time.AfterFunc.
+var timeAfterFunc = func(d time.Duration, f func()) Timer {
+	return time.AfterFunc(d, f)
+}
+
+// Timer is used so we can supply a fake timer during testing.
+type Timer interface {
+	Stop() bool
+}
+
+// Mode controls the pairing policy of the pairing manager.
+type Mode string
+
+const (
+	// ModeUnset is the same as ModeDisabled, but this value prevents the
+	// plan from marshalling the Mode explicitly.
+	ModeUnset Mode = ""
+	// ModeDisabled means no pairing is possible
+	ModeDisabled Mode = "disabled"
+	// ModeSingle means only a single client can pair.
+	ModeSingle Mode = "single"
+	// ModeMultiple means multiple clients can pair.
+	ModeMultiple Mode = "multiple"
+)
+
+type PairingConfig struct {
+	Mode Mode `yaml:"mode,omitempty"`
+}
+
+func (c *PairingConfig) Validate() error {
+	switch c.Mode {
+	case ModeUnset, ModeDisabled, ModeSingle, ModeMultiple:
+	default:
+		return fmt.Errorf("cannot support pairing mode %q: unknown mode", c.Mode)
+	}
+	return nil
+}
+
+// Implements the optional Zeroer interface as used by the YAML library
+// for deciding when to marshal a section or not.
+func (c *PairingConfig) IsZero() bool {
+	return c.Mode == ModeUnset
+}
+
+func (c *PairingConfig) Combine(other *PairingConfig) {
+	if other.Mode != ModeUnset {
+		c.Mode = other.Mode
+	}
+}
+
+type PairingManager struct {
+	state  *state.State
+	mu     sync.Mutex
+	config *PairingConfig
+	open   bool
+	// timer controls the duration of the pairing window.
+	timer Timer
+	// cancelTimerFunc cancels a pairing window if enabled.
+	cancelTimerFunc context.CancelFunc
+}
+
+func NewManager(state *state.State) *PairingManager {
+	m := &PairingManager{
+		state: state,
+		config: &PairingConfig{
+			Mode: ModeDisabled,
```

Comment:
Shouldn't this be `ModeUnset`? Maybe I'm misunderstanding the differences between `ModeUnset` and `ModeDisabled`